### PR TITLE
New module - github_key

### PR DESCRIPTION
**Pull request by rwe**

### CODE REVIEW

File: source_control/github_key.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+      - SSH key name
+    required: true
+  pubkey:
+    description:
+      - SSH public key value. Required when state=present
+    required: false
+  state:
+    description:
+      - Whether to remove a key, ensure that it exists, or update its value.
+    choices: ['present', 'absent', 'updated']
```
Comment (bcoca) [inline source_control/github_key.py]: present should take care of updated

File: source_control/github_key.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = '''
+module: github_key
+short_description: Manage GitHub access keys.
+description:
```
Comment (bcoca) [inline source_control/github_key.py:20]: missing version_added

File: source_control/github_key.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+  name:
+    description:
+      - SSH key name
+    required: true
+  pubkey:
+    description:
+      - SSH public key value. Required when state=present
+    required: false
+  state:
+    description:
```
Comment (bcoca) [inline source_control/github_key.py:38]: missing required: false

File: source_control/github_key.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+      - GitHub Access Token with permission to list and create public keys.
+    required: true
+  name:
+    description:
+      - SSH key name
+    required: true
+  pubkey:
+    description:
+      - SSH public key value. Required when state=present
+    required: false
```
Comment (bcoca) [inline source_control/github_key.py:35]: when required is false, you need to show default

**Comment by bcoca** [comment]
some general issues with not handling any data for check mode nor trying to handle changed status and no error checking on the return, which should then be a fail_json call and not exit_json.

File: source_control/github_key.py (hunk 4 above)
Comment (rwe) [inline source_control/github_key.py:35]: Is that true even if it's a conditionally required parameter? There is no meaningful default value.

File: source_control/github_key.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    description:
+      - GitHub Access Token with permission to list and create public keys.
+    required: true
+  name:
+    description:
+      - SSH key name
+    required: true
+  pubkey:
+    description:
+      - SSH public key value. Required when state=present
```
Comment (rwe) [inline source_control/github_key.py]: This should be `state=present or state=updated`

File: source_control/github_key.py (hunk 3 above)
Comment (rwe) [inline source_control/github_key.py:38]: Isn't that implicit when specifying a default? When could a parameter have a default but still be required?

I just went through other PRs in `ansible-modules-core` where changes of adding redundant `required` to something with `default` were removed. I'm getting mixed signals on what the policy is there.

File: source_control/github_key.py (hunk 1 above)
Comment (rwe) [inline source_control/github_key.py]: There are scenarios you want to ensure a key is present by name, but don't want to update it to a new value if it already exists. Is the expectation that this should overwrite that key every time?

**Comment by rwe** [comment]
> not handling any data for check mode

In `check_mode`, this module does everything except for send the mutating `POST` or `DELETE` changes. It performs all the searches, calculating of what modifications need to happen, etc. What behavior seems to be missing?

> nor trying to handle changed status and no error checking on the return, which should then be a fail_json call and not exit_json.

Good catch, I've added HTTP error handling  in `GitHubSession.request`. `fetch_url` handles the other error types (connection, etc.) and calls `fail_json` appropriately.

File: source_control/github_key.py (hunk 2 above)
Comment (rwe) [inline source_control/github_key.py:20]: Got it. What version should that be? `VERSION` on the root of this repo is currently marked as `1.8.2`, but not sure if that was the last version number or next release number.

Comment (bcoca) [inline source_control/github_key.py:20]: 2.0

Comment (rwe) [inline source_control/github_key.py:20]: Done

File: source_control/github_key.py (hunk 1 above)
Comment (bcoca) [inline source_control/github_key.py]: yes, in the cases that you don't you can add a force=yes|no parameter (look at copy module)

**Comment by bcoca** [comment]
for check_mode you need to return changed=True or False, possibly detailing what changes are required.

File: source_control/github_key.py (hunk 1 above)
Comment (rwe) [inline source_control/github_key.py]: Got it! That makes sense, I'll make the change. Thanks

File: source_control/github_key.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+      - GitHub Access Token with permission to list and create public keys.
+    required: true
+  name:
+    description:
+      - SSH key name
+    required: true
+  pubkey:
+    description:
+      - SSH public key value. Required when state=present or state=updated.
+    required: false
```
Comment (bcoca) [inline source_control/github_key.py:35]: if required: false a default should be present, when required is true it is not (would make no sense)

**Comment by rwe** [comment]
`check_mode` is returning `changed`, see https://github.com/ansible/ansible-modules-extras/pull/692/files#diff-bb556342af3f47c7143fa9278a6e1909R157 and https://github.com/ansible/ansible-modules-extras/pull/692/files#diff-bb556342af3f47c7143fa9278a6e1909R174

**Comment by rwe** [comment]
This should be ready for another look.

File: source_control/github_key.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    if pubkey:
+        pubkey_parts = pubkey.split(' ')
+        # Keys consist of a protocol, the key data, and an optional comment.
+        if len(pubkey_parts) < 2:
+            module.fail_json(msg='"pubkey" parameter has an invalid format')
+
+        # Strip out comment so we can compare to the keys GitHub returns.
+        pubkey = ' '.join(pubkey_parts[:2])
+    elif state == 'present':
+        module.fail_json(msg='"pubkey" is requred when state=present')
```
Comment (mscherer) [inline source_control/github_key.py]: typo, a 'i' is missing in required

File: source_control/github_key.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+      - GitHub Access Token with permission to list and create public keys.
+    required: true
+  name:
+    description:
+      - SSH key name
+    required: true
+  pubkey:
+    description:
+      - SSH public key value. Required when state=present.
+    required: false
```
Comment (resmo) [inline source_control/github_key.py:35]: add  `default: none` or `default: null` here

File: source_control/github_key.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+    description:
+      - GitHub Access Token with permission to list and create public keys.
+    required: true
+  name:
+    description:
+      - SSH key name
+    required: true
+  pubkey:
+    description:
+      - SSH public key value. Required when state=present.
```
Comment (resmo) [inline source_control/github_key.py]: hint: use C(state=present) for better readability

File: source_control/github_key.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+- name: Read SSH public key to authorize
+  shell: cat /home/foo/.ssh/id_rsa.pub
+  register: ssh_pub_key
+
+- name: Authorize key with GitHub
+  local_action:
+    module: github_key
+    name: 'Access Key for Some Machine'
+    token: '{{github_access_token}}'
+    pubkey: '{{ssh_pub_key.stdout}}'
```
Comment (resmo) [inline source_control/github_key.py:65]: your example is good for registering a public key of a _remote_ system. Maybe an example of a local key using lookup would be nice:

```
- name: Authorize local key with GitHub
  local_action:
    module: github_key
    name: 'Access Key of {{ ansible_hostname }}'
    token: '{{ github_access_token }}'
    pubkey: '{{ lookup('file', '~/.ssh/id_rsa.pub') }}'
```

**Comment by gregdek** [comment]
Adding new process. We will be evaluating all new module PRs according to this process, effective immediately.

# 

Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

**Comment by gregdek** [comment]
Looks like this PR needs revision. Please make the suggested revisions, and when you're done, please add a comment with the text "ready_for_review" and we will continue with the review process.  Thanks!

**Comment by gregdek** [comment]
@erydo A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

**Comment by rwe** [comment]
Thanks @gregdek, I will be addressing @resmo's comments—this fell off my radar.

File: source_control/github_key.py (hunk 10 above)
Comment (rwe) [inline source_control/github_key.py:65]: This module is not intended to register a local machine. Doing so would require saving a privileged github access token to that machine first, and there's little or no benefit to going through that process versus just adding your local key by hand.

The module is intended for registering remote systems: servers, build machines and so on, and I think that showing the local example might be confusing. A newcomer may see that example and fail to make the distinction, and perhaps even not realize that they should have separate keys on different machines.

**Comment by rwe** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @erydo for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

**Comment by gundalow** [comment]
See also https://github.com/ansible/ansible-modules-extras/pull/211 which is similar

**Comment by bcoca** [comment]
this overlaps with #211

**Comment by rwe** [comment]
Could we please get some eyes on this to merge in? All previous comments were addressed and this should be good to go…

**Comment by gundalow** [comment]
Note to self: gundalow_review_this

**Comment by gundalow** [comment]
Thank you for this module

PEP8 and Python 3 checks are good.

It's most likely something wrong I'm doing, but I can't get this module to work.

```
    - debug: msg='{{ssh_pub_key.stdout}}'

    - name: Authorize key with GitHub
      local_action:
        module: github_key
        name: 'Access Key for Some Machine'
        token: 'REDACTED'
        pubkey: '{{ssh_pub_key.stdout}}'
        force: false
```

`fatal: [localhost -> localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"force": false, "name": "Access Key for Some Machine", "pubkey": "ssh-rsa REDACTED user@host", "state": "present", "token": "REDACTED"}, "module_name": "github_key"}, "msg": " failed to send request POST to https://api.github.com/user/keys: HTTP Error 422: Unprocessable Entity"}
`

**Comment by rwe** [comment]
@gundalow Hmm, just to be clear, the token you provided was a personal access token via https://github.com/settings/tokens, not actually the string 'REDACTED' ? I just tried this successfully:

``` yml
- hosts: all
  tasks:
    - name : Get machine public SSH key
      shell: cat ./foo_rsa.pub
      register: ssh_pub_key

    - name: Printing Key
      debug:
        msg: '{{ssh_pub_key.stdout}}'

    - name: Authorize key with GitHub
      local_action:
        module: github_key
        name: 'Ansible Test'
        token: '{{github_provisioning_token}}'
        pubkey: '{{ssh_pub_key.stdout}}'
        force: false
```

**Comment by gundalow** [comment]
I gave given the token just "write:public_key" permissions. If I change the token I get an authentication error.

**Comment by rwe** [comment]
@gundalow The token needs both `write:public_key` as well as `read:public_key`. Otherwise it can't determine whether to perform no change or to delete/create the key. See `ensure_key_present`. Does that help?

**Comment by gundalow** [comment]
Even with `admin:public_key` (which sets `write:public_key` and `read:public_key`) this fails for me.

If I use the command line

```
curl -vu gundalow:mytoken https://api.github.com/user/keys
```

I can see my existing SSH keys

If I remove the `read:public_key` permission and retry the curl I get a 404

**Comment by gundalow** [comment]
Urgh, the none obvious error means this pubic part already exists under a key with a different name.

Module is good to merge

**Comment by gundalow** [comment]
shipit

**Comment by gregdek** [comment]
Thanks again to @erydo for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by alikins** [comment]
Shipping it.

**Comment by dfjaimes** [comment]
seems that this is missing return documentation, leading to broken builds

**Pull request merged**